ksonnet-gen/ksonnet/renderer.go

```diff
@@ -0,0 +1,362 @@
+package ksonnet
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	nm "github.com/ksonnet/ksonnet-lib/ksonnet-gen/nodemaker"
+	"github.com/pkg/errors"
+)
+
+// renderer is an item that can be rendered.
+type renderer interface {
+	Render(parent *nm.Object) error
+}
+
+type baseRenderer struct {
+	name        string
+	description string
+	parent      string
+	ref         string
+}
+
+func newBaseRenderer(field Field, parent string) baseRenderer {
+	return baseRenderer{
+		name:        field.Name(),
+		description: field.Description(),
+		parent:      parent,
+		ref:         field.Ref(),
+	}
+}
+
+func (r *baseRenderer) setter() string {
+	return fieldName(r.name, false)
+}
+
+func (r *baseRenderer) mixin() string {
+	return fieldName(r.name, true)
+}
+
+// LiteralFieldRenderer renders a literal field.
+type LiteralFieldRenderer struct {
+	lf         *LiteralField
+	parentName string
+}
+
+// NewLiteralFieldRenderer creates an instance of LiteralField.
+func NewLiteralFieldRenderer(lf *LiteralField, parentName string) *LiteralFieldRenderer {
+	return &LiteralFieldRenderer{
+		lf:         lf,
+		parentName: parentName,
+	}
+}
+
+// Render renders the literal field in the container.
+func (r *LiteralFieldRenderer) Render(container *nm.Object) error {
+	var rndr renderer
+
+	switch ft := r.lf.FieldType(); ft {
+	case "array":
+		rndr = NewArrayRenderer(r.lf, r.parentName)
+	case "object":
+		rndr = NewObjectRenderer(r.lf, r.parentName)
+	case "string", "boolean", "integer", "number":
+		rndr = NewItemRenderer(r.lf, r.parentName)
+	default:
+		return errors.Errorf("unknown literal field type %s", ft)
+	}
+
+	return rndr.Render(container)
+}
+
+// ReferenceRenderer renders a reference field.
+type ReferenceRenderer struct {
+	baseRenderer
+	rf *ReferenceField
+	tl TypeLookup
+}
+
+// NewReferenceRenderer creates an instance of ReferenceRenderer.
+func NewReferenceRenderer(rf *ReferenceField, tl TypeLookup, parent string) *ReferenceRenderer {
+	return &ReferenceRenderer{
+		baseRenderer: newBaseRenderer(rf, parent),
+		tl:           tl,
+		rf:           rf,
+	}
+}
+
+// Render renders the reference in the container.
+func (r *ReferenceRenderer) Render(container *nm.Object) error {
+	name := r.rf.Name()
+	desc := r.rf.Description()
+
+	mo := nm.NewObject()
+	mixinPreamble(mo, r.parent, name)
+
+	ref := r.rf.Ref()
+
+	ty, err := r.tl.Type(ref)
+	if err != nil {
+		return errors.Wrapf(err, "fetch type %s", ref)
+	}
+
+	RenderFields(r.tl, mo, name, ty.Properties())
+
+	formattedName := formatKind(r.rf.Name())
+
+	container.Set(nm.NewKey(formattedName, nm.KeyOptComment(desc)), mo)
+	_ = genTypeAliasEntry(container, name, ref)
```

Comment:
What's the reason that we don't return error on these?

Reply:
Because the error is in consequential. It's being dropped here. If you were to care that it wouldn't work, you could look at the error. I put this in here because I have another tool that actually cares about the error.